tests/simple/testdata/type_deduction.textproto

```diff
@@ -259,3 +259,497 @@ section {
     }
   }
 }
+
+# These are somewhat complex expressions for edge cases in type inference. The
+# result values are not interesting, but the inferred types are.
+section {
+  name: "flexible_type_parameter_assignment"
+  test {
+    name: "list_parameter"
+    expr: "[[], [[]], [[[]]], [[[[]]]]]"
+    check_only: true
+    typed_result {
+      deduced_type {
+        list_type {
+          elem_type {
+            list_type {
+              elem_type {
+                list_type {
+                  elem_type {
+                    list_type {
+                      elem_type {
+                        list_type {
+                          elem_type { dyn {} }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  test {
+    name: "list_parameter_order_independent"
+    expr: "[[[[[]]]], [], [[[]]]]"
+    check_only: true
+    typed_result {
```

Comment:
Do you know if this is a bug that was carried over from Go or Java? I would have expected list(dyn) to be the better type descriptor, though I could see the type of the first list element (with all of its nesting) matching type `dyn` as the element type in the second list. It's probably more accurate to say `list(dyn)` as the result, but I could see the checkers going either way.

Reply:
I confirmed this behavior in go and java. I think this works because the empty list type candidate keeps a free type parameter that can be inferred to be another list type.

```
Source: [[], [[[]]], [[]], [[[[]]]]]
=====>
[
  []~list(list(list(list(dyn)))),
  [
    [
      []~list(list(dyn))
    ]~list(list(list(dyn)))
  ]~list(list(list(list(dyn)))),
  [
    []~list(list(list(dyn)))
  ]~list(list(list(list(dyn)))),
  [
    [
      [
        []~list(dyn)
      ]~list(list(dyn))
    ]~list(list(list(dyn)))
  ]~list(list(list(list(dyn))))
]~list(list(list(list(list(dyn)))))
```


I think this also works:
`[] + [[]] + [[[]]] + [[[[1]]]]`

Reply:
Thank you!